csrc/mmdeploy/backend_ops/torchscript/optimizer/passes/onnx/fuse_select_assign.cpp

```diff
@@ -0,0 +1,148 @@
+      if (tran_perm[0] != 1 || tran_perm[1] != 0) return false;
+    }
+  }
+
+  {
+    // check gather indice
+    Node* gather_inds = values_map[vmap["gather_inds_2"]]->node();
+    auto inds_val = gather_inds->t(Symbol::attr("value"));
+    if (inds_val.dim() != 0) return false;
+    if (inds_val.data_ptr<long>()[0] != 0) return false;
+  }
+
+  {
+    // check slice start
+    Node* slice = values_map[vmap["slice_2"]]->node();
+    auto start_name = slice->inputs()[1]->debugName();
+    auto start_val = params[start_name];
+    if (start_val.dim() != 1) return false;
+    if (start_val.data_ptr<long>()[0] != 0) return false;
+  }
+
+  // create new node
+  auto graph = node->owningGraph();
+  auto z = values_map[vmap["z"]];
+  auto y = values_map[vmap["y"]];
+  auto where_node = graph->create(Symbol::onnx("Where"), {cmp1->output(), z, y});
+  where_node->insertBefore(node);
+  where_node->output()->copyMetadata(node->output());
+  node->output()->replaceAllUsesWith(where_node->output());
+  return true;
+}
+
+void FuseSelectAssign(Block* block, std::unordered_map<std::string, Tensor>& params,
+                      std::unordered_map<std::string, Value*>& vmap, SubgraphMatcher& matcher) {
+  auto graph = block->owningGraph();
+  auto it = block->nodes().begin();
+  while (it != block->nodes().end()) {
+    auto node = *it;
+    ++it;
+    for (auto block : node->blocks()) {
+      FuseSelectAssign(block, params, vmap, matcher);
+    }
+
+    if (matcher.matchesSubgraphFromAnchorNode(node)) {
+      FuseSelectAssign(node, params, vmap, matcher);
+    }
+  }
+}
+
+void FuseSelectAssign(std::shared_ptr<Graph>& graph,
+                      std::unordered_map<std::string, Tensor>& params) {
+  std::string pattern_str = R"IR(
+      graph(%y, %z, %cmp_1, %cmp_2, %start, %axes):
+        %nz_1 = onnx::NonZero(%cmp_1)
+        %trans_1 = onnx::Transpose(%nz_1)
+        %gather_1 = onnx::GatherND(%z, %trans_1)
+        %reshape_1_shape = onnx::Constant()
+        %reshape_1 = onnx::Reshape(%gather_1, %reshape_1_shape)
+        %shape_2 = onnx::Shape(%y)
+        %expand_2 = onnx::Expand(%cmp_2, %shape_2)
+        %nz_2 = onnx::NonZero(%expand_2)
+        %trans_2 = onnx::Transpose(%nz_2)
+        %trans_shape_2 = onnx::Shape(%trans_2)
+        %gather_inds_2 = onnx::Constant()
+        %gather_2 = onnx::Gather(%trans_shape_2, %gather_inds_2)
+        %unsqueeze_2 = onnx::Unsqueeze(%gather_2)
+        %slice_2 = onnx::Slice(%reshape_1, %start, %unsqueeze_2, %axes)
+        %scatter_2 = onnx::ScatterND(%y, %trans_2, %slice_2)
+        return (%scatter_2)
+  )IR";
+
+  Graph pattern;
+  std::unordered_map<std::string, Value*> vmap;
+  torch::jit::parseIR(pattern_str, &pattern, vmap);
+
+  SubgraphMatcher matcher(pattern, MatchAttribute::NO_MATCH);
+  FuseSelectAssign(graph->block(), params, vmap, matcher);
+  torch::jit::EliminateDeadCode(
+      graph->block(), true,
+      torch::jit::DCESideEffectPolicy::ALLOW_DELETING_NODES_WITH_SIDE_EFFECTS);
```

Comment:
Will `ALLOW_DELETING_NODES_WITH_SIDE_EFFECTS` be safe here especially condering there is `DONT_DELETE_NODES_WITH_SIDE_EFFECTS`?

Reply:
This pass is only used in ONNX Graph. ONNX Graph is SSA, theoretically, there are NO nodes with side effects that exist.